pulsar-broker/src/main/java/org/apache/pulsar/compaction/PulsarTopicCompactionService.java

```diff
@@ -106,6 +111,78 @@ public CompletableFuture<Position> getLastCompactedPosition() {
         return CompletableFuture.completedFuture(compactedTopic.getCompactionHorizon().orElse(null));
     }
 
+    @Override
+    public CompletableFuture<Entry> findEntryByPublishTime(long publishTime) {
+        final Predicate<Entry> predicate = entry -> {
+            try {
+                return Commands.parseMessageMetadata(entry.getDataBuffer()).getPublishTime() >= publishTime;
+            } finally {
+                entry.release();
+            }
+        };
+        return findFirstMatchEntry(predicate);
+    }
+
+    @Override
+    public CompletableFuture<Entry> findEntryByEntryIndex(long entryIndex) {
+        final Predicate<Entry> predicate = entry -> {
```

Comment:
How do you handle if the entry index is minus value?

Reply:
It will return the first entry of compacted data, this is also consistent with the method definition, What do you think?